## Machinle Learning by Stanford University and DeepLearning. Ai ##
Machine Learnig Specialization by Andrew Ng.  
Notes by César Muro Cabral.

 Machine learning (ML) is part of applied mathematics and computer science. It uses tools from mathematical disciplines such as probability, statistics, and optimization theory to extract patter from data.

 The main idea behind machine learning is learnig from examples: we prepare a dataset with examples, and a machine learning systems 'learns' from this dataset. In other words, we give the system the input and desired output, and the system tries to figure out how to do the conversion automatically, without asking a human. 
 
 Then we provide a machine learning model with the training model. This process is called training or sometimes fitting. When training is done, we can use the model by asking it.
 
 In ML the difficult work is done by the machine; we need only supervise the training process to make sure that the model is good. To use ML we need to have data. If no data is avaliable, ML is not possible.
 
 In supervised ML we provide the model with features, we supervise or teach the model by showing it examples. We can express ML mathematically as  
 $$ y=g(X) $$
where $ g $ is the function that we want to learn with machine learning, $X$ is the feature matrix in which rows are feature vectors, and $y$ is the target variable: a vector. 
 

**Machine Learning** is the science of getting computers to learn without being explicitly programmed. It is a subfield of IA.  
Machine learning is the field of study that gives computers the hability to learn without being explicitly programmed.  
In general, the more opportunities you give your algorithm to learn the better it will perform.

Note: Sometime in future building machines as intelligent as you or me. This is sometimes called Artificial General Intelligence or AGI. 

The two main types of machine learnin algorithms are:  
* Supervised learning: It is the most used in real world applications,and it has seen the most rapid avancements and innovations.


* Unsupervised learning: Recommender systems. Reinforcement learning.

The machine learning process has six steps:  
  
1 **Business understanding**  
  
2 **Data understanding**: The next step is data understanding. Here, we try to identify the data sources we can
use to solve the problem.    
  
3 **Data preparation**: In this step, we clean the data, transforming it in such a way that it can be used as
input for a machine learning model.    
  
4 **Modeling**: We decide which machine learning model to use and how to make sure
that we get the best out of it. We need to know how we will measure the performance of the models to select the
best one. It’s very likely that in this step, we need to go back and adjust the way we prepare the data. Perhaps we came up with a great feature, so we go back to the data preparation step to write some code to compute that feature. When the code is done, we train the model again to check whether this feature is good. The goal at this step is to produce a model in such a way that it achieves the best predictive performance.  
  
 A better approach for selecting the best model before deploying it is emulating the scenario of going live. We get our complete dataset, take a part out of it, and train the model on the remaining part of the data. When the training is done, we pretend that the held-out dataset is the new, unseen data, and we use it to measure the performance of our models. This part of data is often called the *validation set*, and the process of keeping part of a dataset away and using it to evaluate performance is called *validation*.  
      
5 **Evaluation**: In this step, we check whether the model lives up to expectations. When we set the goal in the business understanding step, we also define the way of establishing whether the goal is achieved. Typically, we do this by looking at an important business metric and making sure that the model moves the metric in the right direction. 
  
How do we select the best model with the best parameters? To do so, we use the same evaluation scheme. We train the models with different parameters on the training data, apply them to the validation data, and then select the model and its parameters based on the best validation results. 
     
6 **Deployment**: The best way to evaluate a model is to battle-test it: roll it out to a fraction of users and
then check whether our business metric changes for these users.    

The process of selecting the best model and the best parameters for the model is
called **model selection**. We can summarize model selection as follows:  
  
1 We split the data into training, validation, and testing parts.  
  
2 We train each model first on the training part and then evaluate it on validation.  
  
3 Each time we train a different model, we record the evaluation results using the
validation part.  
  
4 At the end, we determine which model is the best and test it on the test dataset.

# SUPERVISED MACHINE LEARNING  


Andrew thinks that the 99% of the economic value created by machine learning is through supervised machine learning.  

Supervised machine learning refers to algorithms to learn $ x (input) \rightarrow y(output) $ mappings. 
The key characteristic of supervised learning is that you give your learning algorithm examples to learn from that includes the right answer. The task of the learning algorithm is to produce more right answers.  

A particular type of supervised learning is Regression: Predict a number from infinitely many possible numbers.  

The second type of supervised learning problem is Clasification. Clasification algorithms predict categories which are a small finite set. The terms output classes and output categories are often used interchangeably

The most lucrative form of supervised learning today is probably used in online advertising. Nearly all the large online ad platforms have a learning algorithm that inputs some information about an ad and some information about you and then tries to figure out if you will click on that ad or not. Because by showing you ads they're just slightly more likely to click on, for these large online ad platforms, every click is revenue, this actually drives a lot of revenue for these companies. 

A bit more formally, we can express a supervised machine learning model mathematically as  
$$y\sim g(X),$$  
where $g$ is the function that we want to learn, $X$ is the feature matrix in which rows, and $y$ is the target variable, a vector.

The goal of machine learning is to learn this function $g$ in such a way that when it gets the matrix $X$, the output is close to the vector $y$. In other words, the function $g$ must be able to take in $X$ and produce $y$. The process of learning g is usually called *training* or *fitting*. 

## Unsupervised machine learning

The most widely used form of machine learning is unsupervised learning. Andrew says that unsupervised learning is just super supervised machine learning. It find patterns or structures that might me in this data. We don't provide the right answer.    


An unsupervised algorithm might decided that data can be assigned to two different groups or two different clusters. it is called **clustering algorithm**. It is used, for example, in google news.  

Then, in unsupervised learning data only comes with inputs x, but not output labels y. Algorithm has to find the structure of data. Other two types of unsupervised algorithms are anomaly detection, which find unusual data points, and dimensionality reduciton, which compress data using fewer numbers.



## Linear regression model with one variable

It just means fitting a straight line to our data.  
Another name is univariate linear regression.  
  
Recall that regression models predicts numbers.  

The data used to train the model is named *training set*. Then to predict, we first train the model to learn from the training set and the our model can make predicitons.    
  
Notation:  
$x$: input variable, or input feature.  
$y$: out put variable, or target variable.  
$m$: number of training examples.
  
Then, we create a function $\hat{y}=f(x)$ where $\hat{y}$ is the estimate or the prediciton for $y$. The function $f$ is called the model.  
  
In the case of linear fitting the model (linear) function is:  
$$f(x)=f_{w,b}(x)=wx+b$$  
$w$, and $b$ are the parameters of the model. Recall that $w$ is the slope of the line and $b$ the $y$-axis interception. 
$x$ are the input features.  

## Cost function

It tells us how well the model is doing.  

Im machine learning parameters of the model are the variable you can adjust during training in order to imporve the model. They also are referred as *weights* or *coefficients*.  
  
  The cost function takes the prediction $\hat{y}$ and compares it with the target $y$ by taking the sum of square of the residuals, i.e. $$J(w,b)=\frac{1}{2m}\sum^{m}_{i=1}(y_{i}-\hat{y}_{i})^{2},$$ where we denotate $J(w,b)$ as the cost function, and $\hat{y}_{i}=f_{w,b}(x_{i})$. This is also called *the squared error cost function*.  
  
 We want the parameters such that minimize the cost function $J(w,b)$.

We can visualize $J(w,b)$ as a 3D function with variables $w$ and $b$. A convenient way to visualize a 3D plot is with a Counter Plot.

## Gradient descent

A systematic way to find the values of the parameter that fit the model. There is the algorithm *gradient descent*. It is also used to train deep learning models.   

Gradient descent allows us to minimize any function $J(w_{1},w_{2},...,w_{n},b)$.  
  
Outline:  
* Start with initial guesses of the parameters.  
* Keep changing the parameters to reduce $J$.  
* Until we settle at or near minimum.  
  
  
Then, the algorithm follows the expression (the sign = is in programming context not in rigid math context)
$$tempw=w-\alpha \frac{\partial}{\partial w} J(w,b), $$  
$$tempb=b-\alpha \frac{\partial}{\partial b}J(w,b),$$
$$w=tempw,\quad b=tempb$$
where $\alpha$ is the *learning rate* which is usually a small positive number between 0 and 1. It controls how big of a step you take downhill.  
We simultaneously update $w$ and $b$ and repeat until convergence.  

### Learning rate

The choice of the learning rate, alpha will have a huge impact on the efficiency of your implementation of gradient descent.
And if alpha, the learning rate is chosen poorly rate of descent may not even work at all.  
  
If we take $\alpha$ very very small, then we end up taking small baby steps, and the following step is also minuscule. The result is that we end up decreasing the cost function $J$ but incredoble slowly; Gradient descent may be slow.  
  
If $\alpha$ is too large we get further and further away from the minimum. Gradient descent  fail to converge, diverge

# Regression with multiple input variables

Now we have $n$ features, i.e. $x_{j},\quad j=1,...,n$ features to predict $y$.  

Moreover, $\vec{x}^{\,(i)}$ denotes features of the $i^{th}$ training example.  
  
Now, our linear model takes the form:
$$f_{w,b}(\vec{x})=w_{1}x_{1}+w_{2}x_{2}+...+w_{n}x_{n}+b$$.  
  
We can define the vector $\vec{w}=(w_{1},w_{2},...,w_{n})$. Then $\vec{w}$ and $b$ are the parameters of the model.  
  
Then the model easily reads as $f_{\vec{w},b}=\vec{w}\cdot\vec{x}+b$  

## Vectorization

In Python you can define a vector with the function np.array(), for example np.array([1,2,3,4]).  
The dot product between to vectors $\vec{w}$ and $\vec{x}$ can be written in Python as np.dot(w,x)  
  
  Then, our model takes the form $f=np.dot(w,x)+b$.  
    
 With vectorization the computer acts in parallel. 

Now, the cost funciton is a function of $\vec{w}$, i.e. $J(\vec{w},b)$, and the gradient descent is written as 
$$tempw_{j}=w_{j}-\alpha \frac{\partial}{\partial w_{j}} J(\vec{w},b), $$  
$$tempb=b-\alpha \frac{\partial}{\partial b}J(\vec{w},b),$$
$$w=tempw,\quad b=tempb$$

# Gradient descent in practice

## Feature scaling

Then, how scale features?  
  
* One way is that if we have $x_{min}\leq x \leq x_{max}$ then $x_{scaled}=\frac{x}{x_{max}}$.  
  
* Mean normalization: $x_{scaled}=\frac{x-\mu}{x_{max}-x_{min}}$.  

* Z-score normalization: $x_{scaled}=\frac{x-\mu}{\sigma}$.  
  


## Checking gradient descent for convergence

Making sure that gradient descent is working well. One way is plotting the # of iterations vs J. Then it shoould tends to zero as increasing the numer of iterations. The curve is called *learning curve*.  
  
  If $J$ ever increases after one iteration that means either $\alpha$ is choosen poorly and it usually means $\alpha$ is too large or there could be a bug in the code.
  
  An approach is to performa an automatic convergence test. Let $\epsilon$ a small number, then if $J\leq \epsilon$ declare convergence. 

The learning algorithm will run much better with an appropiate choice of learning rate.  

**Values of $\alpha$ to try**: 0.001, 0.01, 0.1, 1.  

Choose $\alpha$ too small, then big and choose between them

## Feature engineering

Using intuition to design new features, by transforming or combining original features is called feature engineering.  
To improve our model further, we can add more features to the model: we create others and add them to the
existing features. As we already know, this process is called feature engineering.

## Poynomial regression

Now we fit with a polynomial $f_{\vec{w},b}=b+w_{1}x+w_{2}x^{2}+w_{3}x^{3}+...$.  
It is important to apply feature scaling to get the features into comparable ranges of values

# Classification with logistic regression

Class=category

## Logistic regression

We have two categorical variables, let us say 0 and 1.  
We fit a S-shaped curve called the *logistic function* or *sigmoid function* 
$$g(z)=\frac{1}{1+e^{-z}},\quad 0<g(z)<1.$$  
  
  In the logistic regression model $z=\vec{w} \cdot\ \vec{x}+b$, then
 $$g(z=\vec{w} \cdot\ \vec{x}+b)=\frac{1}{1+e^{-(\vec{w} \cdot\ \vec{x}+b)}}$$

Think logist regression function as outputing the probability that the class or the label $y$ will be equal to 1 given a certain input $x$

## Decision boundary

Define the treshold. With the sigmoid function a common choice would be to pick 0.5.  
  
  The model predicts 1 whenever $\vec{w}\cdot \vec{x}+b \geq 0$. Conversely, when $$\vec{w}\cdot \vec{x}+b \leq 0$$ the model predicts $y=0$.  
  The point where $z=\vec{w} \cdot \vec{x}+b=0$

## Cost function for logistic regression

The square error function is not the ideal for logistic regression. Set $m$ as the number of training examples and $n$ the number of features, with the logistic regression model $f_{\vec{w},b}=\frac{1}{1+e^{-(\vec{w} \cdot \vec{x}+b)}}$.   
  
By employing the same function $J(\vec{w},b)=\frac{1}{2m}\sum_{i=1}^{m}(f_{\vec{w},b}(x^{(i)})-y^{(i)})$ the non-linear nature of the model results in a “wiggly”, non-convex cost function with many potential local minima.  

We define the general form of the logistic loss function as 
$$L(f_{\vec{w},b}(\vec{x}^{(i)}),y^{(i)})=-log(f_{\vec{w},b}(\vec{x}^{(i)})) \quad if y^{(1)}=1$$
$$ =-log(1-f_{\vec{w},b}(\vec{x}^{(i)}))\quad if y^{(i)}=0.   $$
The further prediction $f_{\vec{w},b}(\vec{x},b)$ is from target $y^{(i)}$, the higher the loss.  

**Simplified cost function**  
The loss function takes the simplified form 
$$ L(f_{\vec{w},b}(\vec{x}^{(i)}),y^{(i)})=-y^{(i)}log(f_{\vec{w},b}(\vec{x}^{(i)}))-(1-y^{(i)})log(1-f_{\vec{w},b}(\vec{x}^{(i)})), $$ 
  
and the cost function: 
$$J(\vec{w},b)=\frac{1}{m} \sum^{m}_{i=1}[L(f_{\vec{w},b}(\vec{x}^{(i)}),y^{(i)})] $$

**Gradient descent implementation - Training logistic regression**

Then, our lost function is  
$$J(\vec{w},b)=-\frac{1}{m}\sum_{i=1}^{m}[ y^{(i)}log(f_{\vec{w},b}(\vec{x}^{(i)})) +(1-y^{(i)})log(1-f_{\vec{w},b}(\vec{x}^{(i)})) ]  $$  
and the algorithm of gradient descent  
repeat {
$$  w_{j}=w_{j}-\alpha \frac{\partial}{\partial w_{i}} J(\vec{w},b)  $$  
$$ b=b-\alpha \frac{ \partial }{\partial b} J(\vec{w},b)  $$ 
} $j=1,2,...,n$

## The problem of overfitting

When the data does not fit well the problem is called *underfitting* or *high-bias*.    
  
If the learning algorithm has very strong preconception and fits the data perfectly but the result are wiggly curves. When a model fits the training data well but does not work well with new examples that are not in the training set. This is called overfitting. Another terms is that the algorithm has high variance.  
  
We can say that the objective of machine learning is to get a model that is not under fitting or over fitting  

**Regularization to reduce overfitting**   

* One way to overcome this problem is collecting more data.  
  
* Another option is to see if you can use fewer features. Choosing the most appropiate set of features to use is sometimes called *feature selection*.   
  
* *Regularization*: Reduce the size of the parameters $w_{j}$. It prevents the features having an overly large effect   

**Cost function with regularization**  
Adding smaller values for the parameters $w_{1},...,w_{n},b$. We penalize all the $w_{j}$ parameters.  
$$J(\vec{w},b)=\frac{1}{2m}\sum_{i=1}^{m}(f_{\vec{w},b}(\vec{x}^{(i)})-y^{(i)})+\frac{\lambda}{2m} \sum_{j=1}^{n}w_{j}^{2} $$  
$\lambda>0$ is the *regularization parameter*. By convention we are not penalizing the parameter $b$. The parameter $\lambda$ balances keeping $w_{j}$ small and fitting the data. Increasing the regularization parameter lambdalambda reduces overfitting by reducing the size of the parameters.  For some parameters that are near zero, this reduces the effect of the associated features.

## Regularized linear regression  

Then, our cost function for linear regression is  
$$J(\vec{w},b)=\frac{1}{2m}\sum_{i=1}^{m}(f_{\vec{w},b}(\vec{x}^{(i)})-y^{(i)})+\frac{\lambda}{2m} \sum_{j=1}^{n}w_{j}^{2} $$ and we add to the algorithm  
repeat {
$$  w_{j}=w_{j}-\alpha \frac{\partial}{\partial w_{i}} J(\vec{w},b)  $$  
$$ b=b-\alpha \frac{ \partial }{\partial b} J(\vec{w},b)  $$ 
} $j=1,2,...,n$.  
Then, we have an extra term in the $w_{j}$ that reads as $w_{j}\left(1- \alpha \frac{\lambda}{m} \right)$ which decreases the value of $w_{j}$ each iteration by a little bit.  



## Regularized logistic linear regression

The gradient descent update is very similar for this case. We just add the extra term $\frac{\lambda}{2m}\sum_{j=1}^{n}w_{j}^{2}$, i.e.  
$$J(\vec{w},b)=-\frac{1}{m}\sum_{i=1}^{m}[ y^{(i)}log(f_{\vec{w},b}(\vec{x}^{(i)})) +(1-y^{(i)})log(1-f_{\vec{w},b}(\vec{x}^{(i)})) ]+\frac{\lambda}{2m} \sum_{j=1}^{n}w_{j}^{2} $$ 


# ADVANCED ALGORTIHMS - Course 2

Origins: Algorithms that try to mimic the brain.  
  
The first application area of deep learning and had huge impact on was probably speech recognition. Then evolves to computer vision and after text (NLP) and so on.    
  
  Artifical neural networks uses a very simplified mathematical model of a neuron. The "neuron" takes 1 or more inputs , it does some computation, and it ouputs some other number which then could be an input to be a second neuron. In DL we want to simulate many such neurons at the same time. 
      
For certain types of application where we have a lot of data and we are able to train a very large neural network we can take advantage of that huge amount of data the we could attain performance on anything.         

  
Another way to think of a neuron is as a tiny little computer whose only job is to input one number or a few numbers and the to output one number or maybe few other numbers.  
  
We group the neurons in *layers*, which take as inputs the features. A layer can have multiple neurons or a single neuron. The last layer is called output layer. The set of the input feautures is called input layer. Each neuron in the layer has as input every previous feature.  
  
Notation: input layer $\vec{x}$. For the *activion values* we use the notation $\vec{a}$, to finally obtain a scalar $a$. The layer in the middle is called hidden layer.  
  
We can think neural networks as logistic regression that learn its own features.  
 
Summarizing, a neural network takes a input layer which is a vector a features. It is input to a hidden layer which produces another vector with new features called activitions. Then this activions are input for the last layer which produces and scalar named final activation.  

Architecture for a neural network. The numbers of layers, number of neurons per layer, etc.

## Neural network model  - Week 1


**Neural network layer**  
  
The fundamental building block of most modern neural networks is a layer of neurons. 
  
We will emplot the notation with superscripts of the form $[x]$ for layer numbering with number x.  Then, $\vec{a}^{1}$ is the ouyput of layer 1 and the input of layer 2. If the layer 2 only has one neuron which applies the sigmoid function, then the output is the scalar $a_{1}=g(\vec{w}_{1}\cdot \vec{a}_{1}+b_{1})$.  
  
So that's how a neural network works. Every layer inputs a vector of numbers and applies a bunch of logistic regression units to it, and then computes another vector of numbers that then gets passed from layer to layer until you get to the final output layers computation, which is the prediction of the neural network. Then you can either threshold at 0.5 or not to come up with the final prediction.  
  
**More complex neural networks**  
By convetion when we say a neural network has $n$ layers, that includes all the hidden layers and output layer, but we don't count the hidden layer.  
  
Suppose we have 4 layers. Layer 3 inputs a vector $\vec{a}^{[2]}$. The output is a vector $\vec{a}^{3}=[a_1,a_2,a_3]$ where $a_{1}^{[3]}=g(\vec{w}_{1}^{[3]}\cdot \vec{a}^{[2]}+b_{1}^{[3]})$, $a_{2}^{[3]}=g(\vec{w}_{2}^{[3]}\cdot \vec{a}^{[2]}+b_{2}^{[3]})$, ... . In general, the activation of layer $l$ and unit $j$ is $a_{j}^{[l]}=g(w^{[l]}_{j}\cdot \vec{a}^{[l-1]}+\vec{b}^{[l]}_{j})$. In this case the *activation function* is the sigmoid function.  
Recall that $\vec{a}^{[0]}=\vec{x}$. 
  
**Inference: making predictions (forward propagation)**  
When computation goes from left to right throug a serie of layers, the algorithm is called forward propagation, because we are propagating the activitions of the neurons. The network architecture with more hidden units initially and then the number of units decreases as you get closer to the output layer is common. 


**Tensorflow and Keras**  
Tensorflow is a machine learning package developed by Google. In 2019, Google integrated Keras into Tensorflow and released Tensorflow 2.0. Keras is a framework developed independently by François Chollet that creates a simple, layer-centric interface to Tensorflow. This course will be using the Keras interface. 

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf
from tensorflow.keras.layers import Dense, Input
from tensorflow.keras import Sequential
from tensorflow.keras.losses import MeanSquaredError, BinaryCrossentropy
from tensorflow.keras.activations import sigmoid

In [ ]:
X_train = np.array([0., 1, 2, 3, 4, 5], dtype=np.float32).reshape(-1,1)  # 2-D Matrix
Y_train = np.array([0,  0, 0, 1, 1, 1], dtype=np.float32).reshape(-1,1)  # 2-D Matrix

In [ ]:
pos = Y_train == 1
neg = Y_train == 0
X_train[pos]

In [ ]:
fig,ax = plt.subplots(1,1,figsize=(4,3))
ax.scatter(X_train[pos], Y_train[pos], marker='x', s=80, c = 'red', label="y=1")
ax.scatter(X_train[neg], Y_train[neg], marker='o', s=100, label="y=0", facecolors='none', 
              edgecolors="blue",lw=3)

ax.set_ylim(-0.08,1.1)
ax.set_ylabel('y', fontsize=12)
ax.set_xlabel('x', fontsize=12)
ax.set_title('one variable plot')
ax.legend(fontsize=12)
plt.show()

We can implement a 'logistic neuron' by adding a sigmoid activation. This section will create a Tensorflow Model that contains our logistic layer to demonstrate an alternate method of creating models. Tensorflow is most often used to create multi-layer models. The Sequential model is a convenient means of constructing these models.

In [ ]:
model = Sequential(
    [
        tf.keras.layers.Dense(1, input_dim=1,  activation = 'sigmoid', name='L1')
    ]
)

model.summary() shows the layers and number of parameters in the model. There is only one layer in this model and that layer has only one unit. The unit has two parameters, 𝑤 and 𝑏.

In [ ]:
model.summary()

In [ ]:
logistic_layer = model.get_layer('L1')
w,b = logistic_layer.get_weights()
print(w,b)
print(w.shape,b.shape)

In [ ]:
set_w = np.array([[2]])
set_b = np.array([-4.5])
# set_weights takes a list of numpy arrays
logistic_layer.set_weights([set_w, set_b])
print(logistic_layer.get_weights())

In [ ]:
a1 = model.predict(X_train[0].reshape(1,1))
print(a1)

**Tensorflow and Keras implementation - Inference on code**  
Tensorflow stores data in tensors.  
Let us draw the general implementation:  

import tensorflow as tf  
x=np.array([...])  
layer_1=Dense(units=25,activation='sigmoid')  
layer_2=Dense(units=1,activation='sigmoid')  
a1=layer_1(x) 
a2=layer_2(a1)
  
The function Dense() defines the layer, units argument is the number of neurons, and activation is the activation function.  

There are some inconsistencies between how data data is represented in TensorFlow and Numpy.  
  
Tensorflow uses tensors to represent the data, incluiding vectors which must be initialize as matrix, i.e. with 1xn dimensions arrays.    

**Building a neural network architecture**  


In [5]:
from IPython.display import display
import ipywidgets as widgets

button = widgets.Button(description="Mostrar/Cerrar Código")
output = widgets.Output()

def on_button_clicked(_):
    with output:
        display(display_id)
        
button.on_click(on_button_clicked)

display(button)
display_id = display(widgets.HTML('''<h3>Código oculto</h3>'''), display_id=True)

Button(description='Mostrar/Cerrar Código', style=ButtonStyle())

HTML(value='<h3>Código oculto</h3>')

In [8]:
def toggle_cell_code(button_id):
    """
    Adds a button to toggle (show/hide) the code cell but not the output. 
    
    Parameters
    ----------
    button_id : str
        An identifier for cells that will hide/show when button is pressed.
    """
    
    from IPython.display import display_html
    my_html = '''
    <button type="button" id="%s" onclick="code_toggle('%s')">Codigo</button>
    <script>
    function code_toggle(my_id) {
        // get the parent element for the cell code and output
        var p = $("#"+my_id);
        if (p.length==0) return;
        while (!p.hasClass("cell")) {
            p = p.parent();
            if (p.prop("tagName") =="body") return;
        }
    // get the cell code and toggle its value
    var cell_code = p.find(".input");
    cell_code.toggle();
    }
    </script>
    ''' %(button_id, button_id)
    return display_html(my_html, raw=True)

In [9]:
toggle_cell_code("un_string_unico_y_reconocible")

X=np.array([[185.32,  12.69],
       [259.92,  11.87],
       [231.01,  14.41],
       [175.37,  11.72],
       [187.12,  14.13],
       [225.91,  12.1 ],
       [208.41,  14.18],
       [207.08,  14.03],
       [280.6 ,  14.23],
       [202.87,  12.25],
       [196.7 ,  13.54],
       [270.31,  14.6 ],
       [192.95,  15.2 ],
       [213.57,  14.28],
       [164.47,  11.92],
       [177.26,  15.04],
       [241.77,  14.9 ],
       [237.  ,  13.13],
       [219.74,  13.87],
       [266.39,  13.25],
       [270.45,  13.95],
       [261.96,  13.49],
       [243.49,  12.86],
       [220.58,  12.36],
       [163.59,  11.65],
       [244.76,  13.33],
       [271.19,  14.84],
       [201.99,  15.39],
       [229.93,  14.56],
       [204.97,  12.28],
       [173.19,  12.22],
       [231.51,  11.95],
       [152.69,  14.83],
       [163.42,  13.3 ],
       [215.95,  13.98],
       [218.04,  15.25],
       [251.3 ,  13.8 ],
       [233.33,  13.53],
       [280.24,  12.41],
       [243.02,  13.72],
       [155.67,  12.68],
       [275.17,  14.64],
       [151.73,  12.69],
       [151.32,  14.81],
       [164.9 ,  11.73],
       [282.55,  13.28],
       [192.98,  11.7 ],
       [202.6 ,  12.96],
       [220.67,  11.53],
       [169.97,  12.34],
       [209.47,  12.71],
       [232.8 ,  12.64],
       [272.8 ,  15.35],
       [158.02,  12.34],
       [226.01,  14.58],
       [158.64,  12.24],
       [211.66,  14.17],
       [271.95,  14.97],
       [257.16,  11.71],
       [281.85,  13.96],
       [161.63,  12.52],
       [233.8 ,  13.04],
       [210.29,  14.72],
       [261.24,  13.69],
       [256.98,  13.12],
       [281.56,  13.92],
       [280.64,  11.68],
       [269.16,  13.74],
       [246.34,  12.27],
       [224.07,  12.66],
       [164.24,  11.51],
       [272.42,  14.18],
       [177.68,  12.53],
       [212.86,  14.77],
       [165.88,  15.37],
       [277.43,  12.48],
       [236.51,  12.94],
       [244.14,  11.85],
       [213.45,  13.85],
       [234.57,  14.27],
       [270.34,  12.47],
       [170.68,  13.06],
       [226.79,  15.34],
       [245.92,  14.45],
       [281.32,  12.57],
       [185.03,  13.19],
       [189.88,  14.1 ],
       [278.48,  12.11],
       [219.92,  14.21],
       [216.58,  15.15],
       [249.48,  15.03],
       [165.09,  12.28],
       [158.87,  14.82],
       [279.98,  11.56],
       [256.55,  14.41],
       [272.61,  12.58],
       [246.49,  12.45],
       [160.26,  14.48],
       [155.7 ,  14.3 ],
       [188.27,  13.45],
       [270.36,  12.47],
       [213.22,  12.92],
       [175.7 ,  13.39],
       [174.52,  14.7 ],
       [233.  ,  12.63],
       [281.37,  12.88],
       [240.62,  14.43],
       [185.81,  11.55],
       [270.5 ,  15.33],
       [172.98,  12.11],
       [208.41,  13.89],
       [283.51,  15.35],
       [283.36,  12.48],
       [230.85,  13.24],
       [181.24,  11.76],
       [172.78,  12.93],
       [161.88,  12.1 ],
       [156.03,  13.99],
       [216.52,  12.47],
       [221.06,  13.2 ],
       [238.99,  15.23],
       [197.69,  14.08],
       [179.55,  15.26],
       [233.39,  12.13],
       [184.7 ,  12.14],
       [174.18,  12.73],
       [261.11,  13.33],
       [187.42,  13.18],
       [186.1 ,  14.43],
       [157.94,  12.66],
       [193.64,  12.23],
       [249.65,  12.22],
       [190.56,  11.73],
       [252.  ,  12.96],
       [238.55,  12.37],
       [152.94,  12.79],
       [255.17,  14.85],
       [197.09,  14.89],
       [156.8 ,  13.59],
       [184.75,  13.26],
       [179.92,  15.07],
       [190.79,  15.28],
       [164.73,  13.22],
       [209.87,  14.34],
       [196.58,  13.47],
       [159.51,  12.74],
       [247.87,  11.92],
       [212.44,  12.45],
       [172.34,  11.99],
       [259.87,  14.25],
       [201.23,  13.07],
       [248.34,  13.92],
       [273.66,  15.18],
       [215.09,  14.14],
       [223.53,  12.74],
       [211.22,  14.38],
       [224.61,  14.03],
       [215.75,  15.31],
       [254.82,  12.02],
       [259.9 ,  15.17],
       [260.25,  12.87],
       [199.67,  12.47],
       [157.52,  13.39],
       [264.81,  14.58],
       [239.4 ,  14.89],
       [238.98,  12.39],
       [258.43,  12.97],
       [270.16,  12.81],
       [162.41,  14.42],
       [164.53,  14.98],
       [205.61,  14.62],
       [157.1 ,  13.68],
       [241.38,  12.02],
       [232.13,  12.07],
       [191.04,  12.96],
       [233.64,  12.02],
       [174.95,  14.63],
       [246.64,  13.32],
       [188.07,  14.27],
       [213.16,  12.75],
       [268.08,  12.31],
       [258.58,  13.97],
       [237.21,  14.23],
       [251.02,  15.02],
       [274.28,  12.52],
       [172.12,  15.09],
       [177.52,  12.39],
       [258.71,  15.36],
       [264.01,  13.57],
       [200.71,  15.45],
       [249.37,  14.02],
       [151.5 ,  12.28],
       [151.82,  15.13],
       [181.92,  12.18],
       [228.65,  12.31],
       [223.78,  15.3 ],
       [266.63,  12.48],
       [273.68,  13.1 ],
       [220.61,  12.8 ],
       [284.99,  12.73]])

Y=np.array([[1.],
       [0.],
       [0.],
       [0.],
       [1.],
       [1.],
       [0.],
       [0.],
       [0.],
       [1.],
       [1.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [1.],
       [1.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [1.],
       [1.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [1.],
       [1.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.],
       [1.],
       [1.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [1.],
       [1.],
       [0.],
       [0.],
       [0.],
       [1.],
       [1.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [1.],
       [1.],
       [1.],
       [1.],
       [0.],
       [0.],
       [1.],
       [1.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [1.],
       [1.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.]])

Codigo

### Normalize Data
Fitting the weights to the data (back-propagation, covered in next week's lectures) will proceed more quickly if the data is normalized. This is the same procedure you used in Course 1 where features in the data are each normalized to have a similar range. 
The procedure below uses a Keras [normalization layer](https://keras.io/api/layers/preprocessing_layers/numerical/normalization/). It has the following steps:
- create a "Normalization Layer". Note, as applied here, this is not a layer in your model.
- 'adapt' the data. This learns the mean and variance of the data set and saves the values internally.
- normalize the data.  
It is important to apply normalization to any future data that utilizes the learned model.

In [ ]:
print(f"Temperature Max, Min pre normalization: {np.max(X[:,0]):0.2f}, {np.min(X[:,0]):0.2f}")
print(f"Duration    Max, Min pre normalization: {np.max(X[:,1]):0.2f}, {np.min(X[:,1]):0.2f}")
norm_l = tf.keras.layers.Normalization(axis=-1)
norm_l.adapt(X)  # learns mean, variance
Xn = norm_l(X)
print(f"Temperature Max, Min post normalization: {np.max(Xn[:,0]):0.2f}, {np.min(Xn[:,0]):0.2f}")
print(f"Duration    Max, Min post normalization: {np.max(Xn[:,1]):0.2f}, {np.min(Xn[:,1]):0.2f}")

Tile/copy our data to increase the training set size and reduce the number of training epochs.

In [ ]:
Xt = np.tile(Xn,(1000,1))
Yt= np.tile(Y,(1000,1))   
print(Xt.shape, Yt.shape) 

Let's build our model

In [38]:
import tensorflow as tf
tf.random.set_seed(1234)  # applied to achieve consistent results
model = Sequential(
    [
        tf.keras.Input(shape=(2,)),
        Dense(3, activation='sigmoid', name = 'layer1'),
        Dense(1, activation='sigmoid', name = 'layer2')
     ]
)

In [39]:
model.summary()

Model: "sequential_8"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 layer1 (Dense)              (None, 3)                 9         
                                                                 
 layer2 (Dense)              (None, 1)                 4         
                                                                 
Total params: 13
Trainable params: 13
Non-trainable params: 0
_________________________________________________________________


Let's examine the weights and biases Tensorflow has instantiated.  The weights $W$ should be of size (number of features in input, number of units in the layer) while the bias $b$ size should match the number of units in the layer:

In [40]:
W1, b1 = model.get_layer("layer1").get_weights()
W2, b2 = model.get_layer("layer2").get_weights()
print(f"W1{W1.shape}:\n", W1, f"\nb1{b1.shape}:", b1)
print(f"W2{W2.shape}:\n", W2, f"\nb2{b2.shape}:", b2)

W1(2, 3):
 [[ 0.08334005 -0.29660565  0.17884266]
 [-0.56124383 -0.15262699  0.8899205 ]] 
b1(3,): [0. 0. 0.]
W2(3, 1):
 [[-0.4277674 ]
 [-0.88499916]
 [ 0.3626454 ]] 
b2(1,): [0.]


- The `model.compile` statement defines a loss function and specifies a compile optimization.
- The `model.fit` statement runs gradient descent and fits the weights to the data.

In [42]:
# To train this neural network, all you need to do is call to
#functions you need to call model dot compile with some parameters   
model.compile(loss = tf.keras.losses.BinaryCrossentropy(),
    optimizer = tf.keras.optimizers.Adam(learning_rate=0.01))
model.fit(Xt,Yt,epochs=10)

Epoch 1/10
6250/6250 [==============================] - 8s 1ms/step - loss: 0.1783
Epoch 2/10
6250/6250 [==============================] - 8s 1ms/step - loss: 0.1165
Epoch 3/10
6250/6250 [==============================] - 7s 1ms/step - loss: 0.0426
Epoch 4/10
6250/6250 [==============================] - 8s 1ms/step - loss: 0.0159
Epoch 5/10
6250/6250 [==============================] - 8s 1ms/step - loss: 0.0103
Epoch 6/10
6250/6250 [==============================] - 7s 1ms/step - loss: 0.0072
Epoch 7/10
6250/6250 [==============================] - 8s 1ms/step - loss: 0.0052
Epoch 8/10
6250/6250 [==============================] - 7s 1ms/step - loss: 0.0037
Epoch 9/10
6250/6250 [==============================] - 7s 1ms/step - loss: 0.0027
Epoch 10/10
6250/6250 [==============================] - 7s 1ms/step - loss: 0.0021


#### Epochs and batches
In the `compile` statement above, the number of `epochs` was set to 10. This specifies that the entire data set should be applied during training 10 times.  During training, you see output describing the progress of training that looks like this:
```
Epoch 1/10
6250/6250 [==============================] - 6s 910us/step - loss: 0.1782
```
The first line, `Epoch 1/10`, describes which epoch the model is currently running. For efficiency, the training data set is broken into 'batches'. The default size of a batch in Tensorflow is 32. There are 200000 examples in our expanded data set or 6250 batches. The notation on the 2nd line `6250/6250 [====` is describing which batch has been executed.

**Predicitons**  
Once you have a trained model, you can then use it to make predictions. Recall that the output of our model is a probability. In this case, the probability of a good roast. To make a decision, one must apply the probability to a threshold. In this case, we will use 0.5.    
Let's start by creating input data. The model is expecting one or more examples where examples are in the rows of matrix. In this case, we have two features so the matrix will be (m,2) where m is the number of examples.
Recall, we have normalized the input features so we must normalize our test data as well.   
To make a prediction, you apply the `predict` method.  

In [43]:
X_test = np.array([
    [200,13.9],  # postive example
    [200,17]])   # negative example
X_testn = norm_l(X_test)
predictions = model.predict(X_testn)
print("predictions = \n", predictions)

1/1 [==============================] - 0s 53ms/step
predictions = 
 [[9.875393e-01]
 [9.142362e-08]]


To convert the probabilities to a decision, we apply a threshold:

In [44]:
yhat = (predictions >= 0.5).astype(int)
print(f"decisions = \n{yhat}")

decisions = 
[[1]
 [0]]


## General implementation of forward propagation
We construct manually a neural network of two layers using the sigmoid function as the activation function.
We employ the same sets X,y.

**Normalize data**  
We need to normalize the data

In [47]:
print(f"Temperature Max, Min prenormalization: {np.max(X[:,0]):0.2f},{np.min(X[:,0]):0.2f}")
print(f"Duration Max, Min prenormalization: {np.max(Y[:,0]):0.2f},{np.min(Y[:,0]):0.2f}")
norm_l=tf.keras.layers.Normalization(axis=1)
norm_l.adapt(X)
Xn=norm_l(X)
print(f"Temperature Max, Min post normalization: {np.max(Xn[:,0]):0.2f}, {np.min(Xn[:,0]):0.2f}")
print(f"Duration    Max, Min post normalization: {np.max(Xn[:,1]):0.2f}, {np.min(Xn[:,1]):0.2f}")

Temperature Max, Min prenormalization: 284.99,151.32
Duration Max, Min prenormalization: 1.00,0.00
Temperature Max, Min post normalization: 1.66, -1.69
Duration    Max, Min post normalization: 1.79, -1.70


In [48]:
def g(z):
    g=1/(1+np.exp(-z))
    return g
    

We will define the my_dense() function which computes the activations of a dense layer

In [49]:
def my_dense(a_in,W,b):
    units=W.shape[1]
    a_out=np.zeros(units)
    for j in range(units):
        w=W[:,j]
        z=np.dot(w,a_in)+b[j]
        a_out[j]=g(z)
    return(a_out)

The following cell builds a two-layer neural network utilizing the `my_dense` subroutine above.

In [50]:
def my_sequential(x, W1, b1, W2, b2):
    a1 = my_dense(x,  W1, b1)
    a2 = my_dense(a1, W2, b2)
    return(a2)

We use the trained weights and biases from the code in Tensorflow

In [51]:
W1_tmp = np.array( [[-8.93,  0.29, 12.9 ], [-0.1,  -7.32, 10.81]] )
b1_tmp = np.array( [-9.82, -9.28,  0.96] )
W2_tmp = np.array( [[-31.18], [-27.59], [-32.56]] )
b2_tmp = np.array( [15.41] )

Lets make some predictions using our layers and data of the roasted coffe. We define the function matrix.  

In [53]:
def my_predict(X, W1, b1, W2, b2):
    m = X.shape[0]
    p = np.zeros((m,1))
    for i in range(m):
        p[i,0] = my_sequential(X[i], W1, b1, W2, b2)
    return(p)

In [54]:
X_tst = np.array([
    [200,13.9],  # postive example
    [200,17]])   # negative example
X_tstn = norm_l(X_tst)  # remember to normalize
predictions = my_predict(X_tstn, W1_tmp, b1_tmp, W2_tmp, b2_tmp)

To convert probabilites to a decision, we apply a treshold

In [58]:
yhat = (predictions>=0.5).astype(int)
print(f"decisions = \n{yhat}")

decisions = 
[[1]
 [0]]


**Neural networks are vectorized**  
The function my_dense() is a matrix multiplication. The neural network layers are matrix multplications.   
  
If we define X,W, and B as matrix (with np arrays), then the code becomes: 
  
def dense(A_in,W,B):  
    Z=np.matmul(A_in,W)+B  
    A_out=g(z)  
    return A_out  
    
We can also write the matmul function as Z= A @ B.  


## Neural network training  - Week 2
  
Let us show how is the basic form of the code for a neural network of two layers  

In [3]:
n_cols=1
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
model=Sequential()
model.add(Dense(25,activation='sigmoid',input_shape=(n_cols,)))
model.add(Dense(15,activation='sigmoid'))
# The output layer
model.add(Dense(2,activation='softmax'))
# To compile the model we need to specify the loss function
from tensorflow.keras.losses import BinaryCrossentropy
model.compile(loss='BinaryCrossentropy',optimizer='adam')
#The last part is fitting the model
#model.fit(x,y,epochs=10,valid_split_0.3)


Epoch: Number of steps in gradient descent.  
Activation: Function to inference.  

**Model training steps**  


1.- Create the model  
2.a Loss and cost functions. For binary classification we already now the logistic loss function $L(f(\vec{x},y)=-ylog(f(\vec{x}))-(1-y)log(1-f(\vec{x}))$, which is also called *binary cross entropy*. In tensorflow we have the syntax  
from tensorflow.keras.losses import BinaryCrossentropy    
model.compile(loss=BinaryCrossentropy())  
2.b In case of regression (predicting numbers): loss=MeanSquaredError()  
3. Gradient descent. Tensorflow compute derivatives using backward propagation.

### Alternatives to sigmoid activations

Our neural network can be more powerful by choosing different acitvation functions.  
With sigmoid function, whe choose $g(z)=g(\vec{w}\cdot\vec{x}+b)$, now we take $g(x)=max(z,0)$ which is known as ReLu function. The linear activation function $g(z)=z$.  

**Choosing activation function**  
For the output layer, depends the type of variable we want to predict, the Andrew's criteria:  
* Binary classification: Sigmoid function    
* Regression y=+-: Linear activation function    
* Regression y=+: ReLu activation function  
  
For the hidden layers, the most common acitvation function is the ReLu function since is not flat for positive values.  
  
Note: don't use linear activation in the hidden layers of the neural networks, it is likewise performing linear regression.    

## Multiclass clasification

Classification problems with more of tow possible output labels; the target variable y can take on more than two possible values.

The **softmax regression algorithm** is a generalization of logistic regression from binary classification to multiclass classification contexts. Suposse we have N target values $y=1,2,...,N$ (possible outputs), the softmax reads as   
$$a_{j}=P(y=j,\vec{x})=\frac{e^{z_{j}}}{ e^{z_{1}}+e^{z_{2}}+e^{z_{3}}+e^{z_{4}}+\dots+e^{z_{N}}  },\quad j=1,2,\dots,n$$  
 where $a_{1}+a_{2}+\dots+a_{n}=1$.  
 The loss function, knowns as loss entropy function, is:  
 $$loss(a_{1},\dots,a_{n})=-log a_{i}, \quad if y=i$$

Then, we employ the softmax in the last ouput. The code reads as:  
  
import tensorflow as tf    
from tensorflow.keras.models import Sequential    
from tensorflow.keras.layers import Dense 
from tensorflow.keras.losses import SparseCategoricalCrossentropy  
model=Sequential()  
model.add(Dense(25,activation='relu'))  
model.add(Dense(10,activation='relu'))  
model.add(Dense(10,activation='softmax'))  
model.compile(loss=SparseCategoricalCrossentropy())  
    

For multiclass classification, the recommended way to implement softmax regression is to set from_logits=True in the loss function 'linear'.  
When you set from_logits=True, then it expects the output layer of the model to be 'linear' (logits), because the loss function calculates the softmax itself with a more numerically stable method.


import tensorflow as tf    
from tensorflow.keras.models import Sequential    
from tensorflow.keras.layers import Dense 
from tensorflow.keras.losses import SparseCategoricalCrossentropy  
model=Sequential()  
model.add(Dense(25,activation='relu'))  
model.add(Dense(10,activation='relu'))  
model.add(Dense(10,activation='linear'))  
model.compile(loss=SparseCategoricalCrossentropy(from_logits=True))

Moreover, for a multiclass classfication we place in the output layer the same number of neurons as the number of target labels.  

## Advanced optimization

There are better algorithms to minimize the cost function than gradient descent and allow to train the neural network in a better way. The optimizer adam algorithm adjusts the learning rate automatically. Adams stands for Adaptive moment estimation. If $w_{j}$ keeps moving in the same direction, increase $\alpha_{j}$. If $\alpha_{j}$  keeps oscillating, reduce $\alpha_{j}$.      
  model.compile(optimizer='adam',loss='categorical_crossentropy')  
  model.compile(optimizer='adam',loss='mean_squared_error')

**Additional layer types**  
A convolutional layer:  Each neuron only looks at part of the previous layer's inputs.  
Multiple convolutional layers are called convolutional neural network.  

## Advices for applying machine learning systems - Week 3

Machine learning diagnostic: A test that you run to gain insight into what is/isnt working with a learning algorithm, to gain guidance into improving its performance 

## Evaluating a model  


In [5]:
import os
os.getcwd()

'C:\\Users\\cmuro'

In the context of machine learning, diagnotic is a test that you run to gain insight into what is/isn’t working with a learning algorithm.

How to evaluate our model?  
* Split the training set into two subsets: 70% training set, 30% test set. 
  
In the train/test procedure for linear regression (for squared error cost)  
* Fit the parameters by minimizing the cost function $J(\vec{w},b)$  
$$ J(\vec{w},b)=\left[  \frac{1}{2_{m_{train}}} \sum^{m_{train}}_{i=1}( f_{\vec{w},b} (\vec{x}^{(i)})-y^{(i)})^{2}+\frac{\lambda}{2m_{train}} \sum_{j=1}^{n}w_{j}^{2}  \right]  $$
* Compute test error  
$$ J_{test}(\vec{w},b)=\frac{1}{2m_{test}}\left[  \sum_{i=1}^{m_{test}}( f_{\vec{w},b}(\vec{x}_{test}^{(i)}) -y_{test}^{(i)})^{2} \right] $$


In the train/test procedure for classification problem  
* Same, split into teo subsets: training set and test set
* Fit the parameters by minimizing the cost funciton
$$ J(\vec{w},b)=- \frac{1}{m_{train}} \sum_{i=1}^{m_{train}}[ y^{(i)}log( f_{\vec{w},b} (\vec{x}^{(i)}) ) -(1-y^{(i)})( log(1-f_{\vec{w},b}(\vec{x}^{i})))  ] +\frac{1}{2m_{train}} \sum_{j=1}^{n}w_{j}^{2} $$
* Compute the test error  
$$ J_{test}(\vec{w},b)=- \frac{1}{m_{test}} \sum_{i=1}^{m_{test}}[ y_{test}^{(i)}log( f_{\vec{w},b} (\vec{x}_{test}^{(i)}) ) -(1-y_{test}^{(i)})( log(1-f_{\vec{w},b}(\vec{x}_{test}^{i})))  ] +\frac{1}{2m_{test}} \sum_{j=1}^{n}w_{j}^{2} $$  
  
When applying machine learning to classification problems, there's actually one other definition of J tests and
J train that is maybe even more commonly used. Which is instead of using the logistic loss to compute the test error and the training error to instead measure what the fraction of the test set, and the fraction of the training set that the algorithm has misclassified.  

* $J_{test}(\vec{w},b)$ is the fraction of the test set that has been misclassified.  
* $J_{train}(\vec{w},b)$ is the fraction of the training set that has been misclassified.  

**Model selection and training/cross validation/test sets**  
* $J_{test}(\vec{w},b)$ is better estimate of how well the model will generalize to new data compared $J_{train}(\vec{w},b)$  
* We use the cross validation subset to check or trust check the validity or really the accuracy of different models. We have also the cost function $J_{cv}(\vec{w},b)$ which should be low.
* It's considered best practice in machine learning that if you have to make decisions about your model, such as fitting parameters or choosing the model architecture, such as neural network architecture or degree of polynomial if you're fitting a linear regression, to make all those decisions only using your training set and your cross-validation set, and to not look at the test set at all while you're still making decisions regarding your learning algorithm.  

## Diagnosing bias and variance

Looking at the variance and bias of learning algorithm gives you very good guidance at what to do next.  

A more systematic way to diagnose or to find out if the algorithm has high bias or high bias is to look the performance of the algorithm on the training set.
A characteristic Q that the algorithm has high variance will be that $J_{cv}$ is much higher than $J_{train}$. When the model is jist right $J_{train}$ and $J_{cv}$ is low. 
* High bias (under fit), $J_{train}$ will be high ($J_{train}\sim J_{cv}$).
* High variance (overfit) $J_{cv}>>J_{train}$ ($J_{train}$ may be low).  
* High bias and high variance $J_{train}$ will be high and $J_{cv}>>J_{train}$

**Regularization and bias/variance**  
Choose a better $\lambda$ in regularization. Recall that $\lambda$ is a parameter that controls how much you trade-off keeping the parameters $w_{j}$ small versus fitting the training data well.  
* High bias, $\lambda$ large, underfit
* High variance, $\lambda$ small, overfit
  
Cross validation helps to choose a good $\lambda$ parameter. 

**Establishing a baseline level of performance**  
When judging if the training error is high is often useful to establish a baseline level of performance.   
* What is the level of error you can reasonably get to get to?
    
One common way to establish a baseline level performance is to measure how well:
* Human level performance
* Compething algorithm performance
* Guess based on experience

**Learning curve**  
They help to understand hoy your learning algorithm is doing as a function of the amount of experience whereby experencie I mean , for e.g. the number of training examples it has.  
  

When we have a small number of training examples is relatively easy to get zero or very small training error but for a large data set is harder to fit all training examples.  
The $J_{cv}$ error curve will be tipycally higher than the training error curve. 
* If a learning algorithm suffers from high bias, getting more training data will not help.
* If the algororithm suffers high variance getting more training data is likely to help. The cross validation error will come down and approach to train error.

**Deciding what to try next revisited**  
* Try to decide if our algorithm has high bias or high variance.
* For high bias: Try getting additional features, try adding polynomial features, try decreasing $\lambda$.
* For high variance: Get more training examples, try smaller sets of features, try increasing $\lambda$.

**Bias/Variance and neural networks**  
* Bias-variance tradeoff.
* Large neural networks are low bias machine
* If it does not doing well on the training set, just use a bigger neural network; more hidden layers, or more hidden units per layer.
* It it is doing well on the training set, then check on the cross validation set. If the answer is not, then we have high variance and we can try get more data and going back to the training set.
* A large neural network will usually do as well or better than a smaller one so long as regularization is chose appropiately. 
* To implement regularization in tensorflow:  
Dense(100,activation='relu',kernel_regularizer=L2(0.01))

**Machine learning development process**  
Iterative loop of ML development:  
* Choose architecture (model,data,etc)
* Implement a train model
* Diagnostics (bias,variance, and error analysis)

**Error analysis**  
By manually examining a set of examples that the algorithm is misclassifying or mislabeling leads to elucidate which changes to the model are more promising to try next.  
  
**Adding data**  
When training machine learning algorithms, it feels like always we wish we had even more data almost all the time.  
* Trying to add more data of all types can be slow and expensive. Instead, add more data of the types where error analysis has indicated it might help.  
* Beyond getting brand new training examples (x,y), another technique: data augmentation.  
* Augmentation: modifying an existing training example to create a new training example.
* Distortion introduced should be representation of type of noise/distortions in the test set.
* Usually does not help to add purely random/meaningless noise to your data.   
* Data synthesis: using artificial data inputs to create a new training example.  
* Enginner the data used by your system: Convential model-centric approach, or Data centric approach.  
  
**Transfer learning: using data from a different task**    
* Use data from a different task but training the same model architecture except for the final ouput layer (in the case of neural network architecture). Only train output layers parametes, or try all the parameters.  
* It must have the same input types.
* You can download from internet preload trained neural networks and we just need to fine tune with our data.  

**Full cycle of machine learning project**  
* Define project, scope project
* Define and collect data to train the machine learning system
* Train the model, error analysis & iterative improvement
* Deploy in production, monitor and mantain system

For deployment, some software enginering is needed.
Emerging field:
* MLOps, machine learning operations: It refers to the practice of how systematically build, deploy, and mantain machine learning systems 

**Fairness, bias and ethics**  
Taking into account ethics when designing a ml model. 

## Descision three model - Week 4

Features are categorical variables. Binary classification. The structure has the form of "trees", where each feature is a node and in each node the algorithm takes a decision. The first node is called *root node*. The other nodes are called *decision nodes* and the final ones are *leaf nodes*

**Learning process**    
* Decide what feature to use at the root node   
* Then which feature to use next...   
* How to choose what feature to split on at each node: Maximize purity.  
* Decision 2: When do you stop splitting?  
   * When a node is 100% one class  
   * When a splitting node will result in the tree exceding a maximum depth  
   * When improvements in purity score are below a treshold  
   * When number of examples in a node is below a treshold  

**Measuring purity**  
* *Entropy* as measure of impurity. Lets $p_{1}$ fraction of population 1 and $p_{0}=1-p_{1}$. More entropy more impurity.  
The entrypy function is defined as   
$$H(p_{1})=-p_{1}log_{2}(p_{1})-p_{0}log_{2}(p_{0}),$$
Summarizing the entropy function is a measure of impurity of a set of data.  

In decision tree learning, the reduction of entropy is called information gain. The entropy depends on the number of examples, so we calculate the entropy of the rot not minus the weighted average. This quantity is called the *information gain*.  
With major information gain then bigger reduction in entropy.   
The general formula of information gain:  
$$ IG=H(p_{1}^{root})-\left( w^{left}H(p_{1}^{left})+w^{right}H(p_{1}^{right})  \right).  $$

**Putting it together**  
* Start with all the learning examples at the root node
* Calculate information gain for all possible features, and pick the one with the highest information gain
* Split dataset according the selected feature and create left and right branches of the tree
* Keeping repeating the splitting process until stopping criteria is met:
   * When a node is 100% one class
   * When splitting a node will result in the tree exceeding a maximum depth
   * Information gain from additional splits is less than treshold
   * When number of examples in a node is below a treshold  
   
The bigger the maximum depth the bigger the decision tree could cause overfitting

**Using one-hot encoding of categorical features**  
What if we have features that can have more of two discrete values? let us say $n$ values.  
one hot encoding: We create $k$ binary features (0 or 1 valued)  

**Continuous valued features**  
Split into continuous features: Consider many different values for a treshold and take the one that results in the best information gain.  
  
**Regression trees**  
Instead of reducing the impurity we reduce the variance.  

### Tree ensembles

One single-tree is very sensitive to changes of the data. One solution is to build a lot of decision trees. The we get them to vote, and the majority votes of the predictions. it makes the algorithm more robust.   
  
**Random forest algorithm**  
It is a tree ensemble algorithm.  
Given training set of size $m$  
For $b=1$ to $B$:  
   Use sampling with remplacement to create a new training set of size $m$
   Train a decision tree on the new dataset
   
The number $B$ is around 100, people recommend any value from 64 to 200. $B$ stands for bag.  
  
Randomizing the feature choice: ath each node, when choosing a feature to use to split, if $n$ features are available, pick a random subset of $k<n$ features and allow the algorithm to only choose from that subset of features. $k=\sqrt{n}$.  
  
**XGBoost**  
Today by far the most commonly used way or implementation of decision tree ensembles or decision trees there's an album called XGBoost. It runs quickly, the open source implementations are easily used, has also been used very successfully to win many machine learning competitions as well as in many commercial applications. 
Given training set of size $m$  
For $b=1$ to $B$:  
   Use sampling with remplacement to create a new training set of size $m$. But instead of picking from all examples with equal (1/m) probability, make it more likely to pick missclassified examples from previous trained trees.     
   Train a decision tree on the new dataset  
  
Classification:  
from xgboost import XGBClassifier  
model=XGBClassifier()  
model.fit(X_train,y_train)    
y_pred=model.predict(X_test)    
   
Regression:    
from xgboost import XGBRegressor  
model=XGBRegressor()  
model.fit(X_train,y_train)  
y_pred=model.predict(X_test)  

### When to use decision trees  

**Decision trees vs Neural Networks**  
Decision trees and tree ensembles:  
* Works well on structured data  
* Not recommended for unestructured data (images, audio)  
* Fast
* Small trees may be human interpretable  
* If you decide to use decision trees or tree ensembles use XGBOOST for most of the applications   
   
    
Neural networks: 
* Works well on all types of data, incluiding tabular (structured) and unstructured data.  
* May be slower than decision tree
* Works with transfer learning
* When building a system of system of multiple models working together, it might be easier to string together multiple neural networks.   


# Unsupervised Learning, Recommenders, Reinforcement Learning

## Unsupervised learning - Week 1

**Clustering algorithm**    
A clustering algortihm looks at the number of data points and automatically finds data points that are related or similar to each other. We don't have labeled data, and ask to algorithm to find the structure of the data.  
  
**K-means intuition**  
- It takes a random guess at where might be the center of the clusters that you might ask to find. Randomly take the points at where might be the centers of the different clusters
- The algorithm repeat two different hings:
   - Assign points to *cluster centroids*
   - Move cluster centroids
- Assign each point of the data to its closest centroid
- Recompute the centroids  
  
**K-means algorithm**  
* Randomly initialized K cluster centroids $\mu_{1},\mu_{2},\dots,\mu_{k}$  
* Repeat :
 - Assign points to cluster centroids
  for $i=1,\dots,m$  
  $c^{(i)}:=$ index from (1 to $K$) of cluster centroid closest to $x^{(i)}$  
  $min_{k}||x^{(i)}-\mu_{k}||^{2}$
 - Move cluster centroids 
  for i=1,$\dots$ K  
  $\mu_{k}$ average of points assigned to cluster k

**Optimization objective**  
Also optimizing a specific cost function:  
$c^{(i)}$ index of cluster $(1,2,\dots,K)$ to which example $x^{(i)}$ is currently assigned.  
$\mu_{k}$ = cluster centroid $k$.  
$\mu_{c^{(i)}}$ = cluster centroid of cluster to which example $x^{(i)}$ has been assigned. 
  
The cost function or also called the distortion function:    
$$J(c^{(1)},\dots,c^{(m)},\mu_{1},\dots,\mu_{k})=\frac{1}{m}\sum_{i=1}^{m}||x^{(i)}-\mu_{c^{(i)}}||^{2}.$$
* Repeat :
 - Assign points to cluster centroids
  for $i=1,\dots,m$  
  $c^{(i)}:=$ index from (1 to $K$) of cluster centroid closest to $x^(i)$  
  $min_{k}||x^{(i)}-\mu_{k}||^{2}$
 - Move cluster centroids 
  for i=1,$\dots$ K  
  $\mu_{k}$ average of points assigned to cluster k

**Initializing K-means**  
The very first step of K-means algorithm is to choose random locations as the initial guesses for cluster centroids $\mu_{1},\dots,\mu_{k}$.  
Random initialization:  
- Choose $k<m$  
for $i=1$ to 100:
- Randomly pick $K$ training examples
- Run K- means. Get $c^{(1)},\dots,c^{(m)}$ ,$\mu_{1},\dots,\mu_{k}$ equal to those $K$ examples. Computer cost function  
$J(c^{(1)},\dots,c^{(m)},\mu_{1},\dots,\mu_{k})$
- Pick the set of clusters that gave lowest cost $J$  
  
**Choosing the number of clusters**  
- The elbow method: We run $K$ means with a variety of values of $K$ and plot the cost function as the number of clusters. Andrew does not use this method because in a lot of application the right $K$ is often ambiguos.  
- Often, you want to get clusters for some later downstream purpose
- Evaluate $K-$means based on how well it performs on the later purpose

## Anomaly detection

Anomaly detection algorithms look at unlabeled dataset of normal events and thereby learns to detect or to raise a red flag for if there is an unusual events.  
The most common way to carry out anomaly detection is through a technique called desnsity estimation.  
If you have a dataset ${x^{1},x^{2},\cdots,x^{m}}$. The first thing is build a model for porbability $p(x)$ of $x$. the for a new $x_{test}$, we compute the probability $p(x_{x})$ and if it is small, i.e. $p(x_{test})<\epsilon$ with $\epsilon$ a small number, we will raise a flag that this could be an anomaly.  
Anomaly detection is used in fraud detection, manufacturing.  
  
In order to apply anomaly detection, we use the normal distribution. Probability of $x$ is determined by a Gaussian with mean $\mu$ and variance $\sigma^{2}$.    
$$p(x) = \frac{1}{\sqrt{2\pi}\sigma}e^{ \frac{-(x-\mu)^{2}}{2\sigma^{2}}  } $$

Maximum likelihood for $\mu$ and $\sigma$:  
$$ \mu=\frac{1}{m}\sum_{i=1}^{m}x^{(i)}, \quad \sigma^{2}=\frac{1}{m}\sum_{i=1}^{m}(x^{(i)}-\mu)^{2} $$

**Anomaly detection algorithm: Density estimation**   
Training set: $\{\vec{x}^{1},\vec{x}^{2},\cdots,\vec{x}^{m}\}$. Each example $\vec{x}^{i}$ has $n$ features. It is a vector $\vec{x}=(x_{1},\cdots,x_{n})$.  
$$p(\vec{x})=p(x_{1};\mu_{1},\sigma_{1}^{2})*p(x_{2};\mu_{2},\sigma_{2}^{2})*\cdots* p(x_{n};\mu_{n},\sigma_{n}^{2})=\Pi_{j=1}^{n}p(x_{j};\mu_{j},\sigma_{j}^{2}) $$  
   
1. Choose n features $x_{1}$ that you think might be indicative of anomalous examples.
2. Fit parameters: $\mu_{1},\cdots,\mu_{n},\sigma_{1}^{2},\cdots,\sigma_{n}^{2}$.
3. Given a new example $x$, compute $p(x)$.   
4. Anomaly if $p(x)<\epsilon$


**Developing and evaluating an anomaly detection system**  
When developing a learning algorithm is much easier if we have a way of evaluating our learning algorithm. 
  
Assume we have labeled data, of anomalous and non-anomalous examples.  
  
1. Fit model $p(x)$ on training set $x^{(1)},x^{(2)},\cdots,x^{(m)}$
2. On a cross validation/test example $x$, predict. If $p(x)< \epsilon$ (anomaly), if $p(x) \geq \epsilon$ normal.
  
Alternative: No test set  

**Anomaly detection vs supervised learning**  
Anomaly detection:  
- Very small number of positive examples ($y=1$). (0-20 is common)  
- Large number of negative (y=0) examples.  
- Many different "types" of anomalies. Hard for any allgorithm to learn from positive examples what anomalies look like; future animalies may look nothing like any of the anomalous examples we have seen so far.  
  
Supervised learning:  
- Large number of positive negative examples.  
- Enough positive examples for algorithm to get sense of whta positive examples are like, future positive examples likely to be similar to one in training set. 
  
**Choosing what features to use**    
Choosing a good choice of features out to be really important. tune the features for anomaly detection.    
Try to make sure the features you give it are more or less Gaussian. Plot the histogram of the feature. Transform the feature to have a more Gaussian $np.log(x)$. $x\rightarrow log(x+C)$  
  
**Error analysis for anomaly detection**  
We want $p(x)$ large for normal examples of $x$ and small for anomalous examples.   
The most common problem:  
$p(x)$ is comparable for nomal and anomalous examples; it is large for both.  
  
The process:  
- Train the model and see what anomalies in the cross validation set the algortihm is failing to detect.
- Then to look at those examples to see if that can inspire the creation of new features that would allow the algorithm to spot. That example takes on unusually large or unusually small values on the new features, so that you can now successfully flag those examples as anomalies. 
  

## Recommender systems

In usual recommender systems, you have some number of users $n_{u}$ as well some number of items $n_{m}$ that you want to recommend to the users. $r(i,j)=1$ if user $j$ has rated movie $i$. $y^{(i,j)}=$ rating given by user $j$ to movie $i$ (defined only if $r(i,j)=1$).  
  
**Using per-item features**  
We have a number $n$ of features.  
For user j: Predict user j's rating for movie j as $w^{(j)}\cdot x^{(i)}+b^{(j)}$. Then  
$w^{(j)},b^{(j)}=$ parameters for user $j$  
$x^{(i)}=$ feature vector for movie $i$
  
For user $j$ and movie $i$, predict rating: $w^{(j)}\cdot x^{(i)}+b^{(j)}$  
$m^{(j)}=$ no. of movies rated by user $j$  
To learn $w^{(j)},b^{(j)}$.  
  
Let us focus on one user $j$. The cost function to minimize is 
$$J(w^{(j)},b^{(j)})= \frac{1}{2m^{(j)}} \sum_{i:r(i,j)=1}(w^{(j)}\cdot x^{(i)}+ b^{(j)}-y^{(i,j)^{2}} )+\frac{\lambda}{2m^{(j)}} \sum_{k=1}^{n}(w_{k}^{(j)})^{2}  $$  
It is convenient to eliminate the constant $m^{(j)}$.  
The goal of a collaborative filtering recommender system is to generate two vectors: For each user, a 'parameter vector' that embodies the movie tastes of a user. For each movie, a feature vector of the same size which embodies some description of the movie. The dot product of the two vectors plus the bias term should produce an estimate of the rating the user might give to that movie.

**Collaborative filtering algorithm**  
How to learn if we don't have features and we have to learn them from the data. Suposse we have the parameters $w^{(j)},b^{(j)}$ for $n_{u}$ users to learn $x^{(i)}$:  
$$ J(x^{(i)})=\frac{1}{2}\sum_{j:r(i,j)=1}(w^{(j)}\cdot x^{(i)}+b^{(j)}-y^{(i,j)})^{2} +\frac{\lambda}{2} \sum_{k=1}^{n}(x_{k}^{(i)})^{2} $$
  
To learn $x^{(1)},\dots, x^{n_{m}}$:  
$$ \frac{min}{x^{(1)},\cdots,x^{n_{m}}} J(x^{(1)},\dots, x^{n_{m}}) = \frac{1}{2} \sum_{i=1}^{n_{m}} \sum_{j:r(i,j)=1}(w^{(j)}\cdot x^{(i)}+b^{(j)}-y^{(i,j)})^{2}+\frac{\lambda}{2}  \sum_{i=1}^{n_{m}}\sum_{k=1}^{n}(x_{k}^{(i)})^{2}$$
i.e. we minimize $x^{(1)},\dots, x^{n_{m}}$, instead of minimizing  $w^{(1)},\dots, w^{n_{u}},b^{(n_{u})}$ as in the case of learning these parameters. By putting them together both minimization process 
  $$  J(w,b,x)=\frac{1}{2}\sum_{(i,j):r(i,j)=1} (w^{(j)}\cdot x^{(i)} + b^{(j)} -y^{(i,j)}) + \frac{\lambda}{2}\sum_{j=1}^{n_{u}}\sum_{k=1}^{n}(w_{k}^{(j)})^{2}+\frac{\lambda}{2}\sum_{j=1}^{n_{m}}\sum_{k=1}^{n}(x_{k}^{(i)})^{2} $$

Then, the gradient descent algorithm takes the form
repeat {
$$  w^{(j)}_{i}=w^{(j)}_{i}-\alpha \frac{\partial}{\partial w^{(j)}_{i}} J(\vec{w},b)  $$  
$$ b^{(j)}=b^{(j)}-\alpha \frac{ \partial }{\partial b^{(j)}} J(\vec{w},b)  $$ 
$$ x^{(i)}_{k}=x^{(i)}_{k}-\alpha \frac{\partial}{\partial x^{(i)}_{k}}J(w,b,x) $$} i.e. the parameters are $w,b,x$.  

**Binary labels: favs, likes and clicks**  
Previously: Predict $y^{(i,j)}$ as $w^{(j)}\cdot x^{(i)}+b^{(j)}$  
For binary lables, predict the porbability of $y^{(i,j)}=1$ is given by $g(w^{(j)}\cdot x^{(i)}+b^{(j)})$ where $g(z)=\frac{1}{1+e^{-z}}$.   
  
The loss for binary labels: $y^{(i,j)}:\quad f_{(w,b,x)}(x)= g(w^{(j)}\cdot x^{(i)}+b^{(j)})$.  
The binary loss function for a single example:  
$$ L(f_{(w,x,b)}(x),y^{(i,j)})=-y^{(i,j)}log(f_{(w,b,x)}(x))-(1-y^{(i,j)})log(1-f_{w,b,x}(x)) $$
Then, the cost function for collaborative filtering:  
$$  J(w,x,b)=\sum_{(i,j):r(i,j)}L(f_{(w,b,x)}(x),y^{(i,j)}).  $$

### Recommender systems implementation detail

**Mean normalization**  
The algorithm could run better with mean normalization. In the case of the movies, if we normalize the movie ratings to have a consistent average value.  
We take as a matrix the values of the ratings of the users. In each row (each movies) compute the average rating that was given. Then we create the vector $\mu$ with the means. We take the original matrix and substract from every rating the mean rating that it was given. Using this matrix with $y^{(i,j)}$, we can learn $w^{(j)},b^{(j)},x^{(j)}$. For user $j$, on movie $i$ predict: $w^{(j)}\cdot x^{(i)}+b^{(j)}+\mu_{i}$.

### Tensorflow implementation of collaborative filtering  

In [1]:
# Custom training loop
import tensorflow as tf
import numpy as np

In [4]:
# Overview of the program
w=tf.Variable(3.0)
x=1.0
y=1.0
alpha=0.01
iterations=30
from iter in range(iterations):
    # Use Tensroflow's Gradient tape to record the steps
    # used to compute the cost J, to enable auto differentation
    with tf.GradientTape() as tape:
        fwb=w*b
        costJ=(fwb-y)**2
    
    # Use the gradient tape to calculate the gradient
    # of the cost with respect to the parameter w
    [dJdw] = tape.gradient(costJ,[w])
    
    # Run one step of gradient descent by updating
    # the value of w to reduce the cost
    w.assign(-alpha*dJdw)

SyntaxError: invalid syntax (2130464582.py, line 6)

**Finding related items**  
Collaborative filtering algorithm gives you a nice way to find related items.  
The features $x^{(i)}$ of item $i$ are quite hard to interpret, but collectively these features do convey something about what that movie is like. It turns out that given features $x^{(i)}$ of item $i$ we can try to find the item $k$ with $x^{(k)}$ similar to $x^{(i)}$. In particular, given a feature vector $x^{(k)}$, the way we determine what are known as with the samllest distance
$$ \sum_{l=1}^{n}(x_{l}^{(k)}-x_{l}^{(i)})^{2}  $$  
   
Limitations of collaborative filtering: 
* It is not vey good at the cold start problem. How rank new items that few users have rated? How to show something reasonable to new users who have rated few items?

### Content-based filtering  


**Collaborative filtering vs Content-based filtering**  
  
- Collaborative filtering: Recommend items to you based on ratings of users who gave similar ratings as you.  
- Content based filtering: Recommend items to you based on features of user and item to find good match.  
  
 Some user features $x_{\mu}^{(j)}$ for user $j$: Age, Gender, Country, Movies watched, Average rating per genre.  
   
 Movie features $x_{m}^{(i)}$ for movie $i$: Year, Genres, Reviews, Average rating.  
   
 These vectors usually have different sizes.  Learning to match user and movies. Predict rating of user $j$ on movie $i$ as:  
 $$ v_{u}^{(j)}\cdot v_{m}^{(i)} $$  
 where $v_{u}^{(j)}$ is computed from $x_{u}^{(j)}$, and $v_{m}^{(i)}$ from $x_{m}^{(i)}$.  

**Deep learning for content based filtering**  
We have the user and feature networks from which we obtain $v_{u}$ and $v_{m}$, and $g(v_{u}^{(j)}\cdot v_{m}^{(i)})$ to predict the probabilti that $y^{(i,j)}$ is 1.  
We can combine both neural networks to make the prediciton. The cost function  
$$ J=\sum_{(i,j):r(i,j)=1}(v_{u}^{(j)}\cdot v_{m}^{(i)}-y^{(i,j)})^{2}+NN \quad regularization\quad term  $$

To find movies similar to movie $i$: $|| v_{m}^{(k)}-v_{m}^{(i)} ||^{2}$ is small.  

**Recommending from a large catalogue**  
Two steps: Retrieval and Ranking steps.  
Retrieval:  
* Generate large list of plausible item candidates
* Combine retrieved items into list, removing duplicates and items already watched  
  
Retrieving more items results in better perfomance, but slower recommendations. To analyse/optimize the trade-off, carry out offline experiments to see if retrieving additional items results in more relevant reccommendations (i.e., $p(y^{(i,j)})$=1  of items displayed to user are higher )   
      
Ranking:  
* Take list retrieved and rank using learned model
* Display ranked items to user

**Tensorflow Implementation**  
We have to define both networks for the user and the items. Recall that both needs to have the sampe number of neurons at the output. The genral syntax takes the form:  
user_NN=tf.keras.models.Sequential([tf.keras.layers.Dense(256,activation='relu'),  
tf.keras.layers.Dense(128,activation='relu'),  
tf.keras.layers.Dense(32)])  
  
item_NN=tf.keras.models.Sequential([  
tf.keras.layers.Dense(256,activation='relu'),  
tf.keras.layers.Dense(128,activation='relu'),  
tf.keras.layers.Dense(32)  
])    
  
% Create the user input and point to the base network
input_user.tf.keras.layers.Input(shape=(num_user_features))  
vu=user_NN(input_user)  
vu=tf.linalg.12_normalize(vu,axis=1)  
  
% Create the item input and point to the base netwrok
input_item=tf.keras.layers.Input(shape=(num_item_features))  
vm=item_NN(input_item)  
vm=tf.linalg.12_normalize(vm,axis=1)  
  
% measure the similarity of the tow vector outputs  
output=tf.keras.layers.Dot(axes=1)([vu,vm])  
  
% specify the inputs and output of the model    
model=Model([input_user,input_item],output)  
  
% specify the cost function 
cost_fn=tf.keras.losses.MeanSquaredError()  

## Week 3 - Reinforcement learning   
We have the state $s$. the task is to find a function that maps from the state $s$ to an action $a$. Using a neural netwrok using supervised learning to directly learn the mapping from the states $x$ to an action $y$.  
  
A key input in reinforcement learning is something called the *reward   function* which tells when it is doing well or bad, i.e. positive reward or negative reward.  

**The return in reinforcement learning**  
The action through different states. How do you know if a particular kind of rewards is better or work than different set of rewards.  
The return is the sum of the rewards but wighted by one additional factor which is called discount factor which is a number a little bit less than 1.  
$$Return=R_{1}+\gamma R_{2}+\gamma^{2}R_{3}+\dots$$  
what the *discount factor* $\gamma$ does it has the effect of the reinforcement learning algorithm a little bit impatient. The return depends on the actions you take.

In reinforcement learning, our goal is to come up with a function which is called a *policy* $\pi$. It is a function $\pi(s)=a$  that tells you what action to take in every state $s$ so as to maximize the return.    
**Markov decision process** The term Markov in the MDP or
Markov decision process refers to that the future only depends on the current state and not on anything
that might have occurred prior to getting to the current state. In other words, in a Markov decision process,
the future depends only on where you are now, not on how you got here. One other way to think of the Markov decision process formalism is that we have a robot or some other agent that we wish to control and what we get to do is choose actions a and based on those actions, something will happen in the world or in the environment, such as our position in the world changes or we get to sample a piece of rock and execute the science mission. The way we choose the action a is with a policy Pi and based on what happens in the world, we then get to see or we observe back what state we're in, as well as what rewards are that we get. 

**State-action value function (Q-function) definition**  
It is denoted by the letter Q, and it is a function of state as well as the action you want to take in that state.  
$$ Q(s,a)= return $$  
if you start in the state $s$, and take the action $a$ once, then behave optimally after that.  
Picking actions from the states and take the largest value of Qfs. The best possible action in the state $s$ is the action $a$ that gives $max_{a} Q(s,a)$.      
Computing $Q(s,a)$ is an important part of the reinforcement learning algorithm.  
  
**Bellman equations**  
How you can compute these values $Q(s,a)$?. The Bellman equation helps us.  
$s$ the current state, $R(s)$ the reward of the current state. $a$ is the current action. $s'$ the state you get after taking action $a$. $a'$ action that you take in $s'$. the Bellman equation is  
$$Q(s,a)=R(s)+\gamma_{max_{a'}}Q(s',a')$$  


#### Continuous state spaces  
The vector state $s=[x,y,\theta,\dot{x},\dot(y),\dot{\theta}]$ and they can take any value within is valid range.  
We are going to train a neural network to compute or to approximate the state action value function Q(s,a) and that in turn will let us pick good actions.  
The inputs of the neural network is the current state and current action and computes or approximate $Q(s,a)$.  
In a state $s$, use a neural network to compute $Q(s,nothing)$, $Q(s,left)$, $Q(s,main)$, $Q(s,right)$. Pick the action tha maximizes $Q(s,a)$. 
Recall that $Q(s,a)=R(s)+\gamma max_{a'}Q(s',a')$.  
The input data can takes the form $x^{(1)}=(s^{(1)},a^{(1)})$ with the output $y^{(1)}$. Then $y^{(1)}=R(s^{(1)})+\gamma max_{a'}Q(s^{'(1)},a').$  

Learning algorithm:  
Initialize neural network randomly as guess of $Q(s,a)$.
Repeat{Take actions in the lunar lander. Get $(s,a,R(s),s')$. Store 10,000 most recent ($s,a,R(s),s'$) tuples}.  
  
Train neural network:  
Create a training set of 10,000 examples using  
$$x=(s,a)\quad y=R(s)+\gamma max_{a'}Q(s',a') $$  
Train $Q_{new}$ such that $Q_{new}(s,a)\sim y$.  
  
It turns out that there's a change to neural network architecture that make this algorithm much more efficient. Instead, it turns out to be more efficient
to train a single neural network to output all four (in the case of the luna lander) of these values simultaneously.  Now the output unit has four output units, and the job of the neural network is to have the four output units output Q of s, nothing, Q of s, left, Q of s, main, and q of s, right. The job of the neural network is to compute simultaneously the Q value for all four possible actions for when we are in the state s. This turns out to be more efficient because given the state s we can run inference just once and get all four of these values, and then very quickly pick the action a that maximizes Q(s,a).       

**Algorithm refinement: $\epsilon$-greedy policy**  
How to choose action while still learning?  
In some state $s$:  
Option 1: 
  Pick the action $a$ that maximizes $Q(s,a)$.  
Option 2: 
  With probability 0.95, pick the action a that maximizes $Q(s,a)$. Greedy action or Explotation step.
  With probability 0.05, pick and action $a$ randomly. "Exploration".  
$\epsilon-greedy$ policy ($\epsilon=0.05$).    